Framework/Geometry/test/BeamlineRayTracerTest.h

```diff
@@ -0,0 +1,615 @@
+    TS_ASSERT_EQUALS(results.size(), 2);
+
+    // Check they are actually what we expect: 1 with the sample and 1 with the
+    // central detector
+    IComponent_const_sptr centralPixel =
+        m_testInstrument->getComponentByName("pixel-(0;0)");
+    IComponent_const_sptr sampleComp = m_testInstrument->getSample();
+
+    if (!sampleComp) {
+      TS_FAIL("Test instrument has been changed, the sample has been removed. "
+              "Ray tracing tests need to be updated.");
+      return;
+    }
+    if (!centralPixel) {
+      TS_FAIL("Test instrument has been changed, the instrument config has "
+              "changed. Ray tracing tests need to be updated.");
+      return;
+    }
+
+    // Iterate through the results
+    Links::const_iterator resultItr = results.begin();
+
+    // First intersection
+    Link firstIntersect = *resultItr;
+
+    // Based on our test ray, the first intersection should occur at these
+    // distances.
+    TS_ASSERT_DELTA(firstIntersect.distFromStart, 10.001, 1e-6);
+    TS_ASSERT_DELTA(firstIntersect.distInsideObject, 0.002, 1e-6);
+
+    // These should be the correct entry point coordinates
+    TS_ASSERT_DELTA(firstIntersect.entryPoint.X(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(firstIntersect.entryPoint.Y(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(firstIntersect.entryPoint.Z(), -0.001, 1e-6);
+
+    // These should be the correct exit point coordinates
+    TS_ASSERT_DELTA(firstIntersect.exitPoint.X(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(firstIntersect.exitPoint.Y(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(firstIntersect.exitPoint.Z(), 0.001, 1e-6);
+
+    // Component that should have been intersected
+    TS_ASSERT_EQUALS(firstIntersect.componentID, sampleComp->getComponentID());
+
+    // Second intersection
+    ++resultItr;
+    Link secondIntersect = *resultItr;
+
+    // Based on our test ray, the second intersection should occur at these
+    // distances.
+    TS_ASSERT_DELTA(secondIntersect.distFromStart, 15.004, 1e-6);
+    TS_ASSERT_DELTA(secondIntersect.distInsideObject, 0.008, 1e-6);
+
+    // These should be the correct entry point coordinates
+    TS_ASSERT_DELTA(secondIntersect.entryPoint.X(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(secondIntersect.entryPoint.Y(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(secondIntersect.entryPoint.Z(), 4.996, 1e-6);
+
+    // These should be the correct exit point coordinates
+    TS_ASSERT_DELTA(secondIntersect.exitPoint.X(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(secondIntersect.exitPoint.Y(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(secondIntersect.exitPoint.Z(), 5.004, 1e-6);
+
+    // Component that should have been intersected
+    TS_ASSERT_EQUALS(secondIntersect.componentID,
+                     centralPixel->getComponentID());
+  }
+
+  void
+  test_that_a_ray_which_just_intersects_one_component_gives_this_component_only() {
+    // Create the test objects to use
+    create_instrument_and_componentInfo();
+
+    // Vector for a ray
+    V3D testDir(0.010, 0.0, 15.004);
```

Comment:
So the important note here is the the x offset is > the source radius? i.e. the source intersection is not counted.

Reply:
I think so, Bhuvan wrote this bit originally.